src/sctools/reader.py

```diff
@@ -0,0 +1,122 @@
+import os
+import gzip
+import bz2
+from copy import copy
+from collections.abc import Iterable
+
+
+class Reader:
+    """
+    Basic reader object that seamlessly loops over multiple input files
+
+    Can be subclassed to create readers for specific file types (fastq, gtf, etc.)
+    """
+
+    def __init__(self, files='-', mode='r', header_comment_char=None):
+        """
+        :param list|str files: file or list of files to be read. Defaults to sys.stdin
+        :param mode: (Default 'r') returns string objects. Change to 'rb' to
+          return bytes objects.
+        """
+
+        if isinstance(files, str):
+            self._files = [files]
+        elif isinstance(files, Iterable):  # test items of iterable
+            files = list(files)
+            if all(isinstance(f, str) for f in files):
+                self._files = files
+            else:
+                raise TypeError('all passed files must be type str')
+        else:
+            raise TypeError('files must be a string filename or a list of such names.')
+
+        # set open mode:
+        if mode not in {'r', 'rb'}:
+            raise ValueError('mode must be one of r, rb')
+        self._mode = mode
+
+        if isinstance(header_comment_char, str) and mode == 'rb':
+            self._header_comment_char = header_comment_char.encode()
+        else:
+            self._header_comment_char = header_comment_char
+
+    @property
+    def filenames(self):
+        return self._files
+
+    def __len__(self):
+        """
+        return the length of the Reader object. this should typically not be used with sys.stdin,
+        as it will consume the input
+        """
+        return sum(1 for _ in self)
```

Comment:
Would be helpful to clarify that this gives the sum of line counts for all files in Reader and requires fully reading all files.